### PR TITLE
Decouple DAG from DagBag and DAG from SerializedDagModel - remove cyclic imports

### PR DESCRIPTION
**Pull request by mik-laj**
DAG objects is created by DagBag object. DAG object creates a DAGBag instance to load DAGs objects.

---
Issue link: WILL BE INSERTED BY [boring-cyborg](https://github.com/kaxil/boring-cyborg)

Make sure to mark the boxes below before creating PR: [x]

- [X] Description above provides context of the change
- [X] Unit tests coverage for changes (not needed for documentation changes)
- [X] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [X] Relevant documentation is updated including usage instructions.
- [X] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

File: airflow/api/common/experimental/__init__.py [hunk 1]
```diff
@@ -21,7 +21,8 @@
 
 from airflow.configuration import conf
 from airflow.exceptions import DagNotFound, DagRunNotFound, TaskNotFound
-from airflow.models import DagBag, DagModel, DagRun
+from airflow.models import DagModel, DagRun
```
Comment (turbaszek) [inline airflow/api/common/experimental/__init__.py:24]: Should we use explicit imports? 
```suggestion
from airflow.models.dag import DagModel
from airflow.models.dag_run import DagRun
```

File: airflow/utils/cycle_tester.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+def test_cycle(dag):
+    """
+    Check to see if there are any cycles in the DAG. Returns False if no cycle found,
+    otherwise raises exception.
+    """
+    def _test_cycle_helper(visit_map, task_id):
+        """
+        Checks if a cycle exists from the input task using DFS traversal
+        """
+        # print('Inspecting %s' % task_id)
```
Comment (turbaszek) [inline airflow/utils/cycle_tester.py:36]:
```suggestion

```

File: airflow/utils/source_utils.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Comment (turbaszek) [inline airflow/utils/source_utils.py:1]: This is probably only my opinion but `source_utils` is too ambiguous for me. Maybe `code_utils`?

**Review by turbaszek — APPROVED** [review]

File: airflow/models/dag.py [hunk 4]
```diff
@@ -1850,31 +1562,27 @@ def create_dagrun(self,
...
-        dag_models = session.query(DagModel).filter(DagModel.dag_id.in_(dag_ids)).all()
-        try:
-            for dag_model in dag_models:
-                dag_model.is_paused = is_paused
-            session.commit()
-        except Exception:
-            session.rollback()
-            raise
+            filter_query.append(
+                DagModel.root_dag_id.in_(self.dag_id)
```
Comment (mik-laj) [inline airflow/models/dag.py]:
```suggestion
                DagModel.root_dag_id == self.dag_id
```

File: airflow/www/utils.py [hunk 5]
```diff
@@ -457,3 +424,18 @@ def is_utcdatetime(self, col_name):
         return False
 
     filter_converter_class = UtcAwareFilterConverter
+
+
+def get_dag(dag, store_serialized_dags=False):
```
Comment (mik-laj) [inline airflow/www/utils.py:429]: This should probably be deleted.

File: airflow/api/common/experimental/__init__.py (hunk 1 above)
Comment (mik-laj) [reply airflow/api/common/experimental/__init__.py:24]: I prefer not to introduce additional changes in this PR. For now, I focus only on DagBag. This PR is already big.

File: airflow/models/dagbag.py [hunk 6]
```diff
@@ -455,5 +449,5 @@ def sync_to_db(self):
         """
         Save attributes about list of DAG to the DB.
         """
-        from airflow.models.dag import DAG
         DAG.bulk_sync_to_db(self.dags.values())
+        SerializedDagModel.bulk_sync_to_db(self.dags.values())
```
Comment (kaxil) [inline airflow/models/dagbag.py:453]: We don't want to write Serialized DAGs in bulk !

Comment (mik-laj) [reply airflow/models/dagbag.py:453]: The implementation has not changed. Only the location of the method call has changed. Do you have another idea for the name?

Comment (kaxil) [reply airflow/models/dagbag.py:453]: We already have SerializedDagModel.write_dag , can we not use that?

Comment (mik-laj) [reply airflow/models/dagbag.py:453]: I didn't want to add a loop to this method. In other places, we may also need to save many items with one call, e.g. tests. I will add that the previous code was also in the method that had the word "bulk" in the name. Bulk does not necessarily mean that this is done with a single database query. Bulk only means that the method accepts a list of elements.

Comment (mik-laj) [reply airflow/models/dagbag.py:453]: What do you think about save_list_to_db?

**Review by potiuk — APPROVED** [review]

File: airflow/utils/dag_cleaner.py [hunk 7]
```diff
@@ -0,0 +1,288 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Comment (BasPH) [inline airflow/utils/dag_cleaner.py:1]: Unsure about the name "dag_cleaner", it doesn't ring a bell with me. What do you think of:
- dag_clearer
- task_clearer
- task_instance_clearer
- clear_tasks
- clear_task_instances

Comment (turbaszek) [reply airflow/utils/dag_cleaner.py:1]: Maybe `dag_utils` in case we will have more things to put there?

Comment (mik-laj) [reply airflow/utils/dag_cleaner.py:1]: I'm afraid we don't want to see more things in this module. This leads to further cyclical imports.

File: airflow/models/dagbag.py (hunk 6 above)
Comment (kaxil) [reply airflow/models/dagbag.py:453]: I am fine with `SerializedDagModel.bulk_sync_to_db` as long as you add some comments to say we don't intend to send bulk requests for serialized dags and the reason

File: airflow/models/dag.py [hunk 8]
```diff
@@ -1850,31 +1562,27 @@ def create_dagrun(self,
...
         :param including_subdags: whether to include the DAG's subdags
-        :param store_serialized_dags: whether to serialize DAGs & store it in DB
         :param session: session
         """
-        dag_ids = [self.dag_id]  # type: List[str]
+        filter_query = [
+            DagModel.dag_id.in_(self.dag_id),
+        ]
         if including_subdags:
-            dag = self.get_dag(store_serialized_dags)
```
Comment (mik-laj) [inline airflow/models/dag.py:1865]: I'm not sure if this behavior is 100% identical, but it seems sensible from a web UIi perspective.

File: airflow/www/utils.py [hunk 9]
```diff
@@ -457,3 +424,18 @@ def is_utcdatetime(self, col_name):
...
+
+    Calling it from UI should set store_serialized_dags = STORE_SERIALIZED_DAGS.
+    There may be a delay for scheduler to write serialized DAG into database,
+    loads from file in this case.
+    FIXME: remove it when webserver does not access to DAG folder in future.
+    """
+    dag = DagBag(
+        dag_folder=dag.fileloc, store_serialized_dags=store_serialized_dags).get_dag(dag.dag_id)
+    if store_serialized_dags and dag is None:
+        dag = DagBag(dag_folder=dag.fileloc, store_serialized_dags=False).get_dag(dag.dag_id)
```
Comment (mik-laj) [inline airflow/www/utils.py:440]: I deleted the recursion to make this code more readable.

File: airflow/utils/dag_cleaner.py (hunk 7 above)
Comment (mik-laj) [reply airflow/utils/dag_cleaner.py:1]: Small modules with one responsibility are a better idea. 
https://medium.com/@severinperez/writing-flexible-code-with-the-single-responsibility-principle-b71c4f3f883f
In Python, you can look at modules the same way you look at classes. This is very often interchangeable.

**Pull request closed without merging**